### ISSUE TITLE
FetchWorkspaceError: rust-analyzer failed to load workspace: "cargo" "--version" failed: The directory name is invalid. (os error 267)

### ISSUE THREAD

**Issue by DanielT** [rust-lang/rust-analyzer#14649]
Currently when I try to run rust-analyzer in vscode on Windows I always get the error
FetchWorkspaceError: rust-analyzer failed to load workspace: "cargo" "--version" failed: The directory name is invalid. (os error 267)
This happens regardless of which projects are in my vscode workspace. It even happens when the vscode workspace is empty!

I am currently using rust-analyzer nighlty
rust-analyzer version: 0.4.1488-standalone (bc78ebd9d 2023-04-23)

I tried with both stable and nightly rust, in case that made any difference - it does not.
rustc 1.69.0 (84c898d65 2023-04-16)
rustc 1.71.0-nightly (8bdcc62cb 2023-04-20)

I find the error message less than helpful. It would be a really good idea to include **which directory** name is supposedly invalid.

When I have a valid rust project in my vscode workspace, cargo check runs just fine, so something else must be the true problem.

**Comment by lnicola** [comment]
Do you have any settings somewhere, or any relevant environment variables?

**Comment by Veykril** [comment]
The directory must be referring to the current working directory we set for the command I presume which is the parent dir of the cargo toml r-a runs against. I'm not sure what could be causing this there

**Comment by lnicola** [comment]
Sometimes non-ASCII characters can cause problems, but I don't think that's the case here.

**Comment by DanielT** [comment]
My directory names are as boring as they could possibly be - all subdirectories of C:\users\daniel\projects

I just found one thing that might be the cause: when I look at the Logs for "Rust Analyzer Client", I see a large dump of json, which contains one setting that I find concerning
`linkedProjects: [ '.\\autosar-data-specification\\Cargo.toml' ],`

This directory did exist previously, but I deleted it. It seems it might have become a linked project globally, but I have no idea where or how that could have been set

**Comment by lnicola** [comment]
Yeah, there was a bug causing that. It should be in a settings.json somewhere.

**Comment by DanielT** [comment]
I just ran a full text search though my home directory. This setting got added to the workspace.json of the default vscode workspace. Now that I have deleted it rust-analyzer works again.

Rather than closing the issue, I would like to suggest two ways rust-analyzer could be improved:
- It should show the directory name in the error message
- it shouldn't fail at all for linked workspaces, that should (at most!) be a warning

With either of those I would not have just wasted an hour or more figuring out this problem.

**Comment by ryancarins** [comment]
Just flagging I had a similar issue with this. Was partway through a refactor to move to workspaces, but then as other things came up when I stashed and moved to working on something else rust-analyzer broke

Similar solution to just remove the non-existent linked project from .vscode/settings.json

Doesn't seem like something of tremendous priority but it would be nice to have either a more meaningful error or as suggested above consider it a warning and try to run as normal

**Comment by ojob** [comment]
Same issue here: I created the project in a subdirectory, then moved it around. As a result, the rust-analyzer failed to load. 

Updating the path in `.vscode/config.json`, key `"rust-analyzer.linkedProjects", solved the issue.

**Comment by jean-philippe-martin** [comment]
Thank you @ojob ! I had the same problem with rust-analyzer complaining about not finding a folder that I had removed, and the error was in the folder you mention. For me it was `.vscode/settings.json`

**Comment by Veykril** [comment]
We should definitely tell the user that the linked projects setting is set when this error occurs.

**Comment by dravenk** [comment]
```
cat .vscode/settings.json 
```
I fixed my issue by modify the path like below.
```
{
  "rust-analyzer.linkedProjects": [
    "./Cargo.toml"
  ]
}
```
